src/core/negcon_rumble.cpp

```diff
@@ -0,0 +1,813 @@
+    {
+      SetAnalogMode(true, false);
+    }
+  }
+}
+
+bool NeGconRumble::DoState(StateWrapper& sw, bool apply_input_state)
+{
+  if (!Controller::DoState(sw, apply_input_state))
+    return false;
+
+  const bool old_analog_mode = m_analog_mode;
+
+  sw.Do(&m_analog_mode);
+  sw.Do(&m_dualshock_enabled);
+  //sw.DoEx(&m_legacy_rumble_unlocked, 44, false);
+  sw.Do(&m_configuration_mode);
+  //sw.Do(&m_command_param);
+  sw.DoEx(&m_status_byte, 55, static_cast<u8>(0x5A));
+
+  u16 button_state = m_button_state;
+  sw.DoEx(&button_state, 44, static_cast<u16>(0xFFFF));
+  if (apply_input_state)
+    m_button_state = button_state;
+  else
+    m_analog_mode = old_analog_mode;
+
+  sw.Do(&m_command);
+
+  sw.DoEx(&m_rumble_config, 45, {0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF});
+  sw.DoEx(&m_rumble_config_large_motor_index, 45, -1);
+  sw.DoEx(&m_rumble_config_small_motor_index, 45, -1);
+  sw.DoEx(&m_analog_toggle_queued, 45, false);
+
+  MotorState motor_state = m_motor_state;
+  sw.Do(&motor_state);
+
+  if (sw.IsReading())
+  {
+    for (u8 i = 0; i < NUM_MOTORS; i++)
+      SetMotorState(i, motor_state[i]);
+
+    if (old_analog_mode != m_analog_mode)
+    {
+        Host::AddIconOSDMessage(fmt::format("Controller{}AnalogMode", m_index), ICON_FA_GAMEPAD,
+                                fmt::format(m_analog_mode ?
+                                              TRANSLATE_FS("AnalogController", "Controller {} switched to analog mode.") :
+                                              TRANSLATE_FS("AnalogController", "Controller {} switched to digital mode."),
+                    m_index + 1u),
+        5.0f);
+    }
+  }
+  return true;
+}
+
+float NeGconRumble::GetBindState(u32 index) const
+{
+  if (index >= static_cast<u32>(Button::Count))
+  {
+    const u32 sub_index = index - static_cast<u32>(Button::Count);
+    if (sub_index >= static_cast<u32>(m_half_axis_state.size()))
+      return 0.0f;
+
+    return static_cast<float>(m_half_axis_state[sub_index]) * (1.0f / 255.0f);
+  }
+  else if (index < static_cast<u32>(Button::Analog))
+  {
+    return static_cast<float>(((m_button_state >> index) & 1u) ^ 1u);
+  }
+  else
+  {
+    return 0.0f;
+  }
+  //if (index == (static_cast<u32>(Button::Count) + static_cast<u32>(HalfAxis::SteeringLeft)) ||
+  //    index == (static_cast<u32>(Button::Count) + static_cast<u32>(HalfAxis::SteeringRight)))
+  //{
+  //  return static_cast<float>(m_half_axis_state[index - static_cast<u32>(Button::Count)]) * (1.0f / 255.0f);
+  //}
+  //else if (index >= static_cast<u32>(Button::Count))
+  //{
+  //  // less one because of the two steering axes
+  //  const u32 sub_index = index - (static_cast<u32>(Button::Analog) + 1);
+  //  if (sub_index >= m_axis_state.size())
+  //    return 0.0f;
+
+  //  return static_cast<float>(m_axis_state[sub_index]) * (1.0f / 255.0f);
+  //}
+  //else
+  //{
+  //  const u32 bit = s_button_indices[index];
+  //  return static_cast<float>(((m_button_state >> bit) & 1u) ^ 1u);
+  //}
```

Comment:
Please don't leave code like this commented out. If it's not needed, remove it. There were a few others below and in the header too.

Reply:
Done.